### PR TITLE
Merge all bn ck nchw branches

### PR DESCRIPTION
**Pull request by bghimireamd**
This merges all the below PRs into one

Backward CK
https://github.com/ROCm/MIOpen/pull/3259
 
Forward training
https://github.com/ROCm/MIOpen/pull/3260
 
Forward Infer
https://github.com/ROCm/MIOpen/pull/3227

### CODE REVIEW

File: src/solver/mha/mha_ck_fa_v2_solver_forward.cpp [hunk 1]
```diff
@@ -239,8 +239,12 @@ MhaCKFlashAttentionV2Forward::GetSolution([[maybe_unused]] const ExecutionContex
...
 
             fmha_runtime_args.p_drop = probability;
-            fmha_runtime_args.drop_seed_offset =
-                std::make_pair(dataFwd.dropoutSeedData, dataFwd.dropoutOffsetData);
+            // fmha_runtime_args.drop_seed_offset =
+            //     std::make_pair(dataFwd.dropoutSeedData, dataFwd.dropoutOffsetData);
+
+            // using dataFwd.dropoutSeedData gpu pointer was causing compiler error
+            // since dropout is disabled for now, placing 0.
+            fmha_runtime_args.drop_seed_offset = std::make_pair(0, 0);
```
Comment (CAHEK7) [inline src/solver/mha/mha_ck_fa_v2_solver_forward.cpp]: @junliume shall we merge that part asap? like that PR https://github.com/ROCm/MIOpen/pull/3333

File: test/bn_spatial_test.cpp [hunk 2]
```diff
@@ -1130,6 +1130,8 @@ struct batch_norm_spatial_driver : test_driver
     batch_norm_spatial_driver()
     {
         this->batch_factor = 4;
+        this->tolerance =
+            4e-3 / std::numeric_limits<T>::epsilon(); // ck solver has tolerance of 4e-3
```
Comment (CAHEK7) [inline test/bn_spatial_test.cpp:1134]: That is actually `33554.4336f` for float and `18014398509481.984` for double, is it intended?

**Review by CAHEK7 — CHANGES_REQUESTED** [review]

File: test/bn_spatial_test.cpp (hunk 2 above)
Comment (bghimireamd) [reply test/bn_spatial_test.cpp:1134]: Since CK kernel have tolerance of 4e-3. Should I just set the tolerance to that?

Comment (CAHEK7) [reply test/bn_spatial_test.cpp:1134]: By the way, I'm not sure that it's an appropriate place for `tolerance` initialization.
For example it is overridden here:
https://github.com/ROCm/MIOpen/blob/42d6e9f4ba44400d0b2164265a2b843b46834334/test/bn_3d_spatial_test.cpp#L1372-L1373

Comment (bghimireamd) [reply test/bn_spatial_test.cpp:1134]: I will directly initialize there (4e-3) then for bn_spatial_test.cpp and bn_3d_spatial_test.cpp

**Review by CAHEK7 — APPROVED** [review]

File: src/include/miopen/batchnorm/problem_description.hpp [hunk 3]
```diff
@@ -52,6 +52,9 @@ struct ProblemDescriptionTag
 {
 };
 
+using index_t                           = int32_t;
+constexpr index_t NumBatchNormReduceDim = 3;
```
Comment (averinevg) [inline src/include/miopen/batchnorm/problem_description.hpp]: Why do we need `index_t`? `NumBatchNormReduceDim` will not exceed `UINT_MAX` in the foreseeable future.

Comment (averinevg) [reply src/include/miopen/batchnorm/problem_description.hpp]: And why do we need `NumBatchNormReduceDim` here?

File: src/include/miopen/batchnorm/problem_description.hpp [hunk 4]
```diff
@@ -307,7 +320,17 @@ struct MIOPEN_INTERNALS_EXPORT ProblemDescription : ProblemDescriptionBase,
...
+    {
+        if(spatial_dim == 2)
+        {
+            return td.GetLayout("NCHW");
+        }
+        else
+        {
+            return td.GetLayout("NCDHW");
+        }
+    }
```
Comment (averinevg) [inline src/include/miopen/batchnorm/problem_description.hpp]: This should not be changed. All the necessary changes we added in https://github.com/ROCm/MIOpen/pull/3213

**Review by averinevg — DISMISSED** [review]

**Review by cderb — APPROVED** [review]

**Review by junliume — APPROVED** [review]

**Pull request merged**